ConfigMapping "Proxy" classes need to be created using the application's classloader

The proxy classes created by ConfigMappingClass need to be created using the same classloader as the Class they are proxying. This is because the generatate proxy class may need to then access other classes in the application. I would also suggest that the generated class should be created in the same package as the original.

The following exception stack is caused by the proxy class being created using the OSGi Bundle ClassLoader for the SmallRye Config bundle but then trying to access the `Location` class which is loaded by the application ClassLoader. The ClassLoader heirarchy does not allow access in that direction.

`Caused by: java.lang.NoClassDefFoundError: org.eclipse.microprofile.config.tck.Location
	at java.base/java.lang.Class.getDeclaredMethodsImpl(Native Method)
	at java.base/java.lang.Class.getDeclaredMethods(Class.java:1134)
	at io.smallrye.config.ConfigMappingInterface.createConfigurationInterface(ConfigMappingInterface.java:554)
	at io.smallrye.config.ConfigMappingInterface$1.computeValue(ConfigMappingInterface.java:49)
	at io.smallrye.config.ConfigMappingInterface$1.computeValue(ConfigMappingInterface.java:47)
	at java.base/java.lang.ClassValue.getFromHashMap(ClassValue.java:226)
	at java.base/java.lang.ClassValue.getFromBackup(ClassValue.java:208)
	at java.base/java.lang.ClassValue.get(ClassValue.java:114)
	at io.smallrye.config.ConfigMappingInterface.getConfigurationInterface(ConfigMappingInterface.java:83)
	at io.smallrye.config.ConfigMappingLoader.getConfigMappingInterface(ConfigMappingLoader.java:31)
	at io.smallrye.config.ConfigMappingProvider.<init>(ConfigMappingProvider.java:70)
	at io.smallrye.config.ConfigMappingProvider$Builder.build(ConfigMappingProvider.java:688)
	at io.smallrye.config.ConfigMappings.registerConfigMappings(ConfigMappings.java:33)
	at io.smallrye.config.inject.ConfigExtension.validate(ConfigExtension.java:175)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:95)
	... 33 more
Caused by: java.lang.ClassNotFoundException: org.eclipse.microprofile.config.tck.Location cannot be found by io.openliberty.io.smallrye.config_1.0.46.202010121521
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:516)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:171)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1081)`

--- a/implementation/src/main/java/io/smallrye/config/ConfigMappingInterface.java
+++ b/implementation/src/main/java/io/smallrye/config/ConfigMappingInterface.java
@@ -65,8 +65,7 @@ final class ConfigMappingInterface implements ConfigMappingMetadata {
     ConfigMappingInterface(final Class<?> interfaceType, final ConfigMappingInterface[] superTypes,
             final Property[] properties) {
         this.interfaceType = interfaceType;
-        this.className = getClass().getPackage().getName() + "." + interfaceType.getSimpleName()
-                + interfaceType.getName().hashCode() + "Impl";
+        this.className = interfaceType.getName() + interfaceType.getName().hashCode() + "Impl";
         this.superTypes = superTypes;
         this.properties = properties;
         this.propertiesByName = Stream.of(properties).collect(toMap(p -> p.getMethod().getName(), p -> p));

--- a/implementation/src/main/java/io/smallrye/config/ConfigMappingLoader.java
+++ b/implementation/src/main/java/io/smallrye/config/ConfigMappingLoader.java
@@ -58,22 +58,22 @@ public final class ConfigMappingLoader {
     @SuppressWarnings("unchecked")
     static <T> Class<? extends ConfigMappingObject> getImplementationClass(Class<T> type) {
         final ConfigMappingMetadata mappingMetadata = ConfigMappingInterface.getConfigurationInterface(type);
-        return (Class<? extends ConfigMappingObject>) loadClass(type.getClassLoader(),
+        return (Class<? extends ConfigMappingObject>) loadClass(type,
                 mappingMetadata.getClassName(),
                 mappingMetadata.getClassBytes());
     }
 
-    static Class<?> loadClass(final ClassLoader classLoader, final String className, final byte[] classBytes) {
+    static Class<?> loadClass(final Class<?> parent, final String className, final byte[] classBytes) {
         // Check if the interface implementation was already loaded. If not we will load it.
         try {
-            return classLoader.loadClass(className);
+            return parent.getClassLoader().loadClass(className);
         } catch (ClassNotFoundException e) {
-            return loadClass(className, classBytes);
+            return defineClass(parent, className, classBytes);
         }
     }
 
-    private static Class<?> loadClass(final String className, final byte[] classBytes) {
-        return ClassDefiner.defineClass(LOOKUP, ConfigMappingLoader.class, className, classBytes);
+    private static Class<?> defineClass(final Class<?> parent, final String className, final byte[] classBytes) {
+        return ClassDefiner.defineClass(LOOKUP, parent, className, classBytes);
     }
 
     private static final class ConfigMappingObjectHolder {

--- a/implementation/src/test/java/io/smallrye/config/ConfigMappingLoaderTest.java
+++ b/implementation/src/test/java/io/smallrye/config/ConfigMappingLoaderTest.java
@@ -26,7 +26,7 @@ class ConfigMappingLoaderTest {
     void loadManually() {
         List<ConfigMappingMetadata> configMappingsMetadata = ConfigMappingLoader.getConfigMappingsMetadata(ServerManual.class);
         configMappingsMetadata.forEach(
-                mappingMetadata -> ConfigMappingLoader.loadClass(ServerManual.class.getClassLoader(),
+                mappingMetadata -> ConfigMappingLoader.loadClass(ServerManual.class,
                         mappingMetadata.getClassName(), mappingMetadata.getClassBytes()));
         ConfigMappingLoader.getImplementationClass(ServerManual.class);
         ConfigMappingLoader.getImplementationClass(ServerManual.class);

